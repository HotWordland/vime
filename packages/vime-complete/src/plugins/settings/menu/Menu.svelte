<svelte:options accessors />

<svelte:window
  on:keydown={onWindowKeyDown}
  on:click="{() => { isActive = false; }}"
/>

<ul
  role="menu"
  tabindex="-1"
  aria-hidden={!isActive}
  {...$$restProps}
  on:click|stopPropagation
  on:keydown
  on:keydown|stopPropagation={onKeyDown}
  on:focus={onMenuOpen}
  bind:this={el}
>
  <slot />
</ul>

<script context="module">
  const Event = {
    OPEN: 'open',
    CLOSE: 'close',
  };
</script>

<script>
  import { tick, createEventDispatcher, onMount } from 'svelte';

  // --------------------------------------------------------------
  // Setup
  // --------------------------------------------------------------

  const dispatch = createEventDispatcher();

  const KeyCode = Object.freeze({
    TAB: 9,
    ENTER: 13,
    ESC: 27,
    SPACE: 32,
    PAGEUP: 33,
    PAGEDOWN: 34,
    END: 35,
    HOME: 36,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
  });

  // --------------------------------------------------------------
  // Props
  // --------------------------------------------------------------

  let el;
  let choices;
  let focusedChoiceIndex = -1;

  export let isActive = false;

  export const getEl = () => el;
  export const getFocusedChoice = () => choices[focusedChoiceIndex];
  export const getFocusedChoiceIndex = () => focusedChoiceIndex;
  export const getChoices = () => choices;
  export const getSubmenu = (choice) => document.getElementById(choice.getAttribute('aria-controls'));
  
  export const focus = () => el.focus();
  
  export const focusChoice = (index) => {
    if (index === -1) {
      focusedChoiceIndex = choices.length - 1;
    } else if (index === choices.length) {
      focusedChoiceIndex = 0;
    } else {
      focusedChoiceIndex = index;
    }
    choices[focusedChoiceIndex].focus();
  };

  export const focusController = () => {
    const controller = document.getElementById($$props['aria-labelledby']);
    if (controller) controller.focus();
  };

  // --------------------------------------------------------------
  // Events
  // --------------------------------------------------------------

  const updateChoices = async () => {
    await tick();
    choices = document.querySelectorAll(`#${$$props.id} > li > button`);
  };

  onMount(updateChoices);

  const onWindowKeyDown = (e) => {
    if (e.keyCode === KeyCode.ESC) isActive = false;
  };

  const onMenuOpen = () => {
    updateChoices();
    focusedChoiceIndex = 0;
    // Prevents forwarding click event that opened the menu to the choice.
    setTimeout(() => { choices[focusedChoiceIndex].focus(); }, 10);
    isActive = true;
  };

  const onMenuClose = () => {
    focusedChoiceIndex = -1;
    focusController();
    isActive = false;
  };

  const onOpenSubmenu = async () => {
    const choice = choices[focusedChoiceIndex];
    choice.click();
    await tick();
    const submenu = getSubmenu(choice);
    if (submenu) submenu.focus();
  };

  const onKeyDown = (e) => {
    const { keyCode } = e;
    const isValidKeyCode = Object.values(KeyCode).includes(keyCode);

    if (isValidKeyCode) { e.preventDefault(); }

    switch (keyCode) {
      case KeyCode.ESC:
        onMenuClose();
        break;
      case KeyCode.DOWN:
      case KeyCode.TAB:
        focusChoice(focusedChoiceIndex + 1);
        break;
      case KeyCode.HOME: case KeyCode.PAGEUP:
        focusChoice(0);
        break;
      case KeyCode.END: case KeyCode.PAGEDOWN:
        focusChoice(choices.length - 1);
        break;
      case KeyCode.UP:
        focusChoice(focusedChoiceIndex - 1);
        break;
      case KeyCode.LEFT:
        onMenuClose();
        break;
      case KeyCode.ENTER: case KeyCode.SPACE:
        onOpenSubmenu();
        break;
      case KeyCode.RIGHT:
        onOpenSubmenu();
        break;
      default:
        break;
    }
  };

  $: dispatch(isActive ? Event.OPEN : Event.CLOSE);
</script>

<style type="text/scss">
  @import '../../../style/common';

  ul {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0;
    padding: 0;
    font-size: $font-size-medium;
    font-weight: $font-weight-regular;
    position: relative;
    overflow: hidden;
    text-align: left;
    white-space: nowrap;
    z-index: 3;
    list-style-type: none;

    &[aria-hidden='true'] {
      display: none;
    }

    &:focus {
      outline: 0;
    }
  }
</style>