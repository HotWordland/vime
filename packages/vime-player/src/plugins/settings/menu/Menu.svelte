<svelte:window
  on:keydown={onWindowKeyDown}
  on:click={dispatchMenuClose}
/>

<ul
  id={$$props.id}
  role="menu"
  tabindex="-1"
  aria-hidden={$$props['aria-hidden']}
  aria-labelledby={$$props['aria-labelledby']}
  on:click|stopPropagation
  on:mousedown|preventDefault
  on:keydown={onKeyDown}
  on:focus={onMenuOpen}
  bind:this={el}
>
  <slot />
</ul>

<script context="module">
  const Event = {
    MENU_CLOSE: 'menuclose',
  };
</script>

<script>
  import { tick, createEventDispatcher } from 'svelte';

  // --------------------------------------------------------------
  // Setup
  // --------------------------------------------------------------

  const dispatch = createEventDispatcher();

  const KeyCode = Object.freeze({
    TAB: 9,
    ENTER: 13,
    ESC: 27,
    SPACE: 32,
    PAGEUP: 33,
    PAGEDOWN: 34,
    END: 35,
    HOME: 36,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
  });

  // --------------------------------------------------------------
  // Props
  // --------------------------------------------------------------

  let el;
  let items;
  let activeItem = 0;

  export const getEl = () => el;
  export const getSubMenu = (item) => document.getElementById(item.getAttribute('aria-controls'));
  
  export const setFocusToItem = (index) => {
    if (index === -1) {
      activeItem = items.length - 1;
    } else if (index === items.length) {
      activeItem = 0;
    } else {
      activeItem = index;
    }
    items[activeItem].focus();
  };

  export const setFocusToController = () => {
    const controller = document.getElementById($$props['aria-labelledby']);
    if (controller) controller.focus();
  };

  // --------------------------------------------------------------
  // Events
  // --------------------------------------------------------------

  const dispatchMenuClose = () => { dispatch(Event.MENU_CLOSE); };
  const onWindowKeyDown = (e) => { if (e.keyCode === KeyCode.ESC) dispatchMenuClose(); };

  const onItemSelect = async () => {
    const item = items[activeItem];
    item.click();
    await tick();
    const subMenu = getSubMenu(item);
    if (subMenu) subMenu.focus();
  };

  const onMenuOpen = async () => {
    await tick();
    items = document.querySelectorAll(`#${$$props.id} > li > button`);
    // Prevents forwarding click event that opened the menu to the item.
    setTimeout(() => { items[activeItem].focus(); }, 10);
  };

  const onMenuClose = () => {
    activeItem = 0;
    dispatchMenuClose();
    setFocusToController();
  };

  const onOpenSubMenu = async () => {
    const item = items[activeItem];
    const subMenu = getSubMenu(item);
    if (!subMenu) return;
    item.click();
    await tick();
    subMenu.focus();
  };

  const onKeyDown = (e) => {
    const { keyCode } = e;
    const isValidKeyCode = Object.values(KeyCode).includes(keyCode);
  
    if (isValidKeyCode) {
      e.preventDefault();
      e.stopPropagation();
    }

    switch (keyCode) {
      case KeyCode.ESC:
        onMenuClose();
        break;
      case KeyCode.DOWN:
        setFocusToItem(activeItem + 1);
        break;
      case KeyCode.TAB:
        onMenuClose();
        break;
      case KeyCode.HOME: case KeyCode.PAGEUP:
        setFocusToItem(0);
        break;
      case KeyCode.END: case KeyCode.PAGEDOWN:
        setFocusToItem(items.length - 1);
        break;
      case KeyCode.UP:
        setFocusToItem(activeItem - 1);
        break;
      case KeyCode.LEFT:
        onMenuClose();
        break;
      case KeyCode.ENTER: case KeyCode.SPACE:
        onItemSelect();
        break;
      case KeyCode.RIGHT:
        onOpenSubMenu();
        break;
      default:
        break;
    }
  };
</script>

<style type="text/scss">
  @import '../../../style/common';

  ul {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0;
    padding: 0;
    font-size: $font-size-medium;
    font-weight: $font-weight-regular;
    position: relative;
    overflow: hidden;
    text-align: left;
    white-space: nowrap;
    z-index: 3;
    list-style-type: none;

    &[aria-hidden='true'] {
      display: none;
    }

    &:focus {
      outline: 0;
    }
  }
</style>