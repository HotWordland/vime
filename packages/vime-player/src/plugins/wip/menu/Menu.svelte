<svelte:window
  on:keydown="{e => { if (e.keyCode === KeyCode.ESC) dispatchMenuClose() }}"
  on:click="{() => dispatchMenuClose()}"
/>

<ul
  id={$$props.id}
  role="menu"
  tabindex="-1"
  aria-hidden={$$props['aria-hidden']}
  aria-labelledby={$$props['aria-labelledby']}
  on:click|stopPropagation
  on:mousedown|preventDefault
  on:keydown={onKeyDown}
  on:focus={onMenuOpen}
  bind:this={ref}
>
  <slot />
</ul>

<script>
  import { tick, createEventDispatcher } from 'svelte'
  import { ctxKey } from '~src/context'

  const dispatch = createEventDispatcher()

  const KeyCode = Object.freeze({
    TAB: 9,
    ENTER: 13,
    ESC: 27,
    SPACE: 32,
    PAGEUP: 33,
    PAGEDOWN: 34,
    END: 35,
    HOME: 36,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40
  })

  let items
  let activeItem = 0
  
  // Two-way binding (readonly).
  export let ref = null

  const dispatchMenuClose = () => dispatch('menuclose')

  const setFocusToItem = index => {
    if (index === -1) index = items.length - 1
    if (index === items.length) index = 0
    activeItem = index
    items[activeItem].focus()
  }

  const setFocusToController = () => {
    const controller = document.getElementById($$props['aria-labelledby'])
    if (controller) controller.focus()
  }

  const getSubMenu = item => document.getElementById(item.getAttribute('aria-controls'))

  const onItemSelect = async () => {
    const item = items[activeItem]
    item.click()
    await tick()
    const subMenu = getSubMenu(item)
    if (subMenu) subMenu.focus()
  }

  const onMenuOpen = async () => {
    await tick()
    items = document.querySelectorAll(`#${$$props.id} > li > button`)
    // Prevents forwarding click event that opened the menu to the item.
    setTimeout(() => { items[activeItem].focus() }, 10)
  }

  const onMenuClose = () => {
    activeItem = 0
    dispatchMenuClose()
    setFocusToController()
  }

  const onOpenSubMenu = async () => {
    const item = items[activeItem]
    const subMenu = getSubMenu(item)
    if (!subMenu) return
    item.click()
    await tick()
    subMenu.focus()
  }

  const onKeyDown = e => {
    const keyCode = e.keyCode
    const isValidKeyCode = Object.values(KeyCode).includes(keyCode)
  
    if (isValidKeyCode) {
      e.preventDefault()
      e.stopPropagation()
    }

    switch (keyCode) {
      case KeyCode.ESC:
        onMenuClose()
        break
      case KeyCode.DOWN:
        setFocusToItem(activeItem + 1)
        break
      case KeyCode.TAB:
        onMenuClose()
        break
      case KeyCode.HOME: case KeyCode.PAGEUP:
        setFocusToItem(0)
        break
      case KeyCode.END: case KeyCode.PAGEDOWN:
        setFocusToItem(items.length - 1)
        break
      case KeyCode.UP:
        setFocusToItem(activeItem - 1)
        break
      case KeyCode.LEFT:
        onMenuClose()
        break
      case KeyCode.ENTER: case KeyCode.SPACE:
        onItemSelect()
        break
      case KeyCode.RIGHT:
        onOpenSubMenu()
        break
    }
  }
</script>

<style type="text/scss">
  @import '../../style/common';

  ul {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0;
    padding: 0;
    font-size: $font-size-medium;
    font-weight: $font-weight-regular;
    position: relative;
    overflow: hidden;
    text-align: left;
    white-space: nowrap;
    z-index: 3;
    list-style-type: none;

    &[aria-hidden='true'] {
      display: none;
    }

    &:focus {
      outline: 0;
    }
  }
</style>