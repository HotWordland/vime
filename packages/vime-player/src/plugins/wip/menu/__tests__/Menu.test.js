import Menu from '../Menu.svelte'

describe('components', () => {
  describe('Menu', () => {
    it('should look structurally correct', () => {
      expect(Menu).toMatchRenderedSnapshot()
    })

    it('should look structurally correct given backButton', () => {
      expect(Menu).toMatchRenderedWithContextSnapshot({
        backButton: {
          title: 'Back Button',
          onClick: () => {}
        }
      })
    })

    it('should look structurally correct given menu items', () => {
      expect(Menu).toMatchRenderedWithContextSnapshot({
        items: [
          {
            title: '#1',
            value: '#1',
            showForwardArrow: true,
            disabled: false,
            ariaControls: 'menu-1',
            ariaExpanded: false,
            onClick: () => {}
          },
          {
            title: '#2',
            value: '#2',
            showForwardArrow: true,
            disabled: false,
            ariaControls: 'menu-2',
            ariaExpanded: false,
            onClick: () => {}
          }
        ]
      })
    })

    it('should look structurally correct given menu radio items', () => {
      expect(Menu).toMatchRenderedWithContextSnapshot({
        backButton: {
          title: 'Back Button',
          onClick: () => {}
        },
        items: [
          { title: '#1', badge: '#1', isChecked: false, onClick: () => {} },
          { title: '#2', badge: '#2', isChecked: false, onClick: () => {} }
        ]
      })
    })
  })
})
