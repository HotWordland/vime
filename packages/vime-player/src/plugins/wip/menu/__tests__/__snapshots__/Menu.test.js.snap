// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Menu should look structurally correct given backButton: toMatchRenderedWithContextSnapshot 1`] = `
<div>
  <ul
    aria-hidden="true"
    class="svelte-198rwkb hidden"
    role="menu"
  >
    <li
      aria-label="Back Button"
      class="svelte-ppdapn back"
      role="menuitem"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-ppdapn"
        >
          <span
            class="svelte-ppdapn"
          >
            Back Button
          </span>
           
          <span
            class="screenReader svelte-ppdapn"
          >
            Go back to previous menu
          </span>
           
        </span>
      </button>
    </li>
     
  </ul>
  
</div>
`;

exports[`components Menu should look structurally correct given menu items: toMatchRenderedWithContextSnapshot 1`] = `
<div>
  <ul
    aria-hidden="true"
    class="svelte-198rwkb hidden"
    role="menu"
  >
     
    
    <li
      aria-controls="menu-1"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="#1"
      class="svelte-ppdapn forward"
      role="menuitem"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-ppdapn"
        >
          <span
            class="svelte-ppdapn"
          >
            #1
          </span>
           
           
          <span
            class="value svelte-ppdapn"
          >
            #1
          </span>
        </span>
      </button>
    </li>
    
    
    <li
      aria-controls="menu-2"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="#2"
      class="svelte-ppdapn forward"
      role="menuitem"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-ppdapn"
        >
          <span
            class="svelte-ppdapn"
          >
            #2
          </span>
           
           
          <span
            class="value svelte-ppdapn"
          >
            #2
          </span>
        </span>
      </button>
    </li>
    
  </ul>
  
</div>
`;

exports[`components Menu should look structurally correct given menu radio items: toMatchRenderedWithContextSnapshot 1`] = `
<div>
  <ul
    aria-hidden="true"
    class="svelte-198rwkb hidden"
    role="menu"
  >
    <li
      aria-label="Back Button"
      class="svelte-ppdapn back"
      role="menuitem"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-ppdapn"
        >
          <span
            class="svelte-ppdapn"
          >
            Back Button
          </span>
           
          <span
            class="screenReader svelte-ppdapn"
          >
            Go back to previous menu
          </span>
           
        </span>
      </button>
    </li>
     
    
    <li
      aria-checked="false"
      aria-label="#1"
      class="svelte-puzrus"
      role="menuitemradio"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-puzrus"
        >
          <span
            class="checkmarkContainer svelte-puzrus"
          >
            <span
              class="checkmark svelte-puzrus"
            />
          </span>
           
          <span
            class="title svelte-puzrus"
          >
            #1
          </span>
           
          <span
            class="badgeContainer svelte-puzrus"
          >
            <span
              class="badge svelte-puzrus"
            >
              #1
            </span>
          </span>
        </span>
      </button>
    </li>
    
    
    <li
      aria-checked="false"
      aria-label="#2"
      class="svelte-puzrus"
      role="menuitemradio"
    >
      <button
        class="control svelte-1b8bwgw"
        type="button"
      >
         
         
        <span
          class="svelte-puzrus"
        >
          <span
            class="checkmarkContainer svelte-puzrus"
          >
            <span
              class="checkmark svelte-puzrus"
            />
          </span>
           
          <span
            class="title svelte-puzrus"
          >
            #2
          </span>
           
          <span
            class="badgeContainer svelte-puzrus"
          >
            <span
              class="badge svelte-puzrus"
            >
              #2
            </span>
          </span>
        </span>
      </button>
    </li>
    
  </ul>
  
</div>
`;

exports[`components Menu should look structurally correct: toMatchRenderedSnapshot 1`] = `
<div>
  <ul
    aria-hidden="true"
    class="svelte-198rwkb hidden"
    role="menu"
  >
     
  </ul>
</div>
`;
